File: homework2/hansika/ques12.java
/*
Technique Used: Linked list fast-slow two-pointer
Time Complexity: O(n)
Space Complexity: O(1)
Time spent: 45 minutes
*/

class TreeNode 
{
    int val;
    LLNode next;
    TreeNode(int v) { this.val = v; }
}

public class ques12 
{

    public static LLNode disconnectCycle(LLNode head) 
    {
        if (head == null || head.next == null) return head;
        LLNode slow = head;
        LLNode fast = head;
        // Detect cycle
        boolean cycle = false;
        while (fast != null && fast.next != null) 
        {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) // meeting point
            {  
                cycle = true;
                break;
            }
        }
        if (!cycle) return head;  // no cycle to break
        //Find the start of the cycle
        slow = head;
        LLNode prev = null;
        while (slow != fast) 
        {
            prev = fast;
            slow = slow.next;
            fast = fast.next;
        }
        // Break the cycle
        prev.next = null;
        return head;
    }


    // Helper print 
    public static void printList(LLNode head) {
        LLNode t = head;
        int count = 0;
        while (t != null && count < 50) //stops after 50 nodes to avoid infinite loop
        {
            System.out.print(t.val);
            if (t.next != null) System.out.print(" -> ");
            t = t.next;
            count++;
        }
        System.out.println();
    }


    public static void main(String[] args) 
    {

        // example1
        LLNode a = new LLNode(10);
        a.next = new LLNode(18);
        a.next.next = new LLNode(12);
        a.next.next.next = new LLNode(9);
        a.next.next.next.next = new LLNode(11);
        a.next.next.next.next.next = new LLNode(4);
        a.next.next.next.next.next.next = a.next.next;

        System.out.println("Input 1 (with cycle): printing limited:");
        printList(a);
        disconnectCycle(a);
        System.out.println("Output 1 (cycle removed):");
        printList(a);
        // eg2
        LLNode b = new LLNode(10);
        b.next = new LLNode(18);
        b.next.next = new LLNode(12);
        b.next.next.next = new LLNode(9);
        b.next.next.next.next = new LLNode(11);
        b.next.next.next.next.next = new LLNode(4);
        b.next.next.next.next.next.next = b.next.next.next.next.next;
        System.out.println("\nInput 2 (with cycle): printing limited:");
        printList(b);
        disconnectCycle(b);
        System.out.println("Output 2 (cycle removed):");
        printList(b);
    }
}

File: homework2/Ques - 12.cpp
// Detect and disconnect a cycle in a singly linked list.
// Method: Floyd's cycle detection to find meeting point, locate cycle start, then sever the link.
// Time: O(n). Space: O(1).

#include <bits/stdc++.h>
using namespace std;

struct Node {
    long long val;
    Node* next;
    Node(long long v) : val(v), next(nullptr) {}
};

Node* buildListWithCycle(const vector<long long>& vals, int pos) {
    // pos: 0-based index where tail will connect; -1 => no cycle
    if (vals.empty()) return nullptr;
    vector<Node*> nodes;
    nodes.reserve(vals.size());
    for (long long v : vals) nodes.push_back(new Node(v));
    for (size_t i = 0; i + 1 < nodes.size(); ++i) nodes[i]->next = nodes[i+1];
    if (pos >= 0 && pos < (int)nodes.size()) nodes.back()->next = nodes[pos];
    return nodes[0];
}

// returns true if a cycle was detected and removed
bool detectAndRemoveCycle(Node* head) {
    if (!head) return false;
    Node* slow = head;
    Node* fast = head;
    bool found = false;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) { found = true; break; }
    }
    if (!found) return false;

    // find start of cycle
    slow = head;
    while (slow != fast) {
        slow = slow->next;
        fast = fast->next;
    }
    Node* start = slow;
    // find the node just before start in the cycle
    Node* ptr = start;
    while (ptr->next != start) ptr = ptr->next;
    // break the cycle
    ptr->next = nullptr;
    return true;
}

void printList(Node* head) {
    if (!head) { cout << "EMPTY"; return; }
    bool first = true;
    for (Node* p = head; p; p = p->next) {
        if (!first) cout << ' ';
        cout << p->val;
        first = false;
    }
}

void freeList(Node* head) {
    while (head) {
        Node* nx = head->next;
        delete head;
        head = nx;
    }
}

int main() {
    ifstream file("input12.txt");
    if (!file) {
        cout << "Error opening file!\n";
        return 1;
    }

    int T;
    if (!(file >> T)) {
        cout << "Invalid input: first token should be the number of test cases.\n";
        return 1;
    }

    for (int tc = 1; tc <= T; ++tc) {
        int m;
        file >> m;
        vector<long long> vals;
        for (int i = 0; i < m; ++i) {
            long long x; file >> x;
            vals.push_back(x);
        }
        int pos;
        file >> pos; // 0-based index for cycle start, -1 = no cycle

        Node* head = buildListWithCycle(vals, pos);
        bool removed = detectAndRemoveCycle(head);

        cout << "Test Case " << tc << ": ";
        printList(head);
        cout << " | " << (removed ? "CYCLE_REMOVED" : "NO_CYCLE") << "\n";

        freeList(head);
    }
    return 0;
}